packages/entity/src/utils/EntityPrivacyUtils.ts

```diff
@@ -0,0 +1,311 @@
+  const entityCompanionProvider = viewerContext.entityCompanionProvider;
+
+  const viewerScopedCompanion = sourceEntity
+    .getViewerContext()
+    .getViewerScopedEntityCompanionForClass(entityClass);
+  {
+    const privacyPolicy = viewerScopedCompanion.entityCompanion.privacyPolicy;
+    const evaluationResult = await asyncResult(
+      privacyPolicy.authorizeDeleteAsync(
+        sourceEntity.getViewerContext(),
+        queryContext,
+        { cascadingDeleteCause },
+        sourceEntity,
+        viewerScopedCompanion.getMetricsAdapter()
+      )
+    );
+    if (!evaluationResult.ok) {
+      if (evaluationResult.reason instanceof EntityNotAuthorizedError) {
+        return false;
+      } else {
+        throw evaluationResult.reason;
+      }
+    }
+  }
+
+  const newCascadingDeleteCause = {
+    entity: sourceEntity,
+    cascadingDeleteCause,
+  };
+
+  // Take entity X which is proposed to be deleted, look at inbound edges (entities that reference X).
+  // These inbound edges are the entities that will either get deleted
+  // or updated with null based on the edge definiton when entity X is deleted.
+  // For each of these inboundEdge entities Y, look at the field(s) on Y that reference X.
+  // For each of the field(s) on Y that reference X,
+  // - if cascade set null, check if user can update Y
+  // - if cascade delete, run canViewerDeleteAsync on Y
+  // Return the conjunction (returning eagerly when false) of all checks recursively.
+
+  const entityConfiguration =
+    viewerScopedCompanion.entityCompanion.entityCompanionDefinition.entityConfiguration;
+  const inboundEdges = entityConfiguration.inboundEdges;
+
+  for (const inboundEdge of inboundEdges) {
+    const configurationForInboundEdge =
+      entityCompanionProvider.getCompanionForEntity(inboundEdge).entityCompanionDefinition
+        .entityConfiguration;
+
+    const loader = viewerContext
+      .getViewerScopedEntityCompanionForClass(inboundEdge)
+      .getLoaderFactory()
+      .forLoad(queryContext, { cascadingDeleteCause: newCascadingDeleteCause });
+
+    for (const [fieldName, fieldDefinition] of configurationForInboundEdge.schema) {
+      const association = fieldDefinition.association;
+      if (!association) {
+        continue;
+      }
+
+      const associatedConfiguration = entityCompanionProvider.getCompanionForEntity(
+        association.associatedEntityClass
+      ).entityCompanionDefinition.entityConfiguration;
+      if (associatedConfiguration !== entityConfiguration) {
+        continue;
+      }
+
+      const entityResultsForInboundEdge = await loader.loadManyByFieldEqualingAsync(
+        fieldName,
+        association.associatedEntityLookupByField
+          ? sourceEntity.getField(association.associatedEntityLookupByField as any)
+          : sourceEntity.getID()
+      );
+
+      const entitiesForInboundEdge = successfulResults(entityResultsForInboundEdge).map(
+        (r) => r.value
+      );
```

Comment:
A subtle point: this will mask system errors like a database outage and programming errors like "undefined is not an object". IMO we should rethrow those so that the calling code gets a system error (which is usually transient) or programming error instead of a permission error.